standalone/examples.go

```diff
@@ -0,0 +1,135 @@
+package standalone
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"github.com/golang/protobuf/jsonpb"
+	"github.com/golang/protobuf/proto"
+	"github.com/mitchellh/mapstructure"
+	"strconv"
+	"strings"
+	"time"
+)
+
+// Example model of an example gRPC request
+type Example struct {
+	Name    string  `json:"name"`
+	Service string  `json:"service"`
+	Method  string  `json:"method"`
+	Request Request `json:"request"`
+}
+
+// MetadataPair (name, value) pair
+type MetadataPair struct {
+	Name  string `json:"name"`
+	Value string `json:"value"`
+}
+
+// Request gRPC request
+type Request struct {
+	Timeout  time.Duration
+	Metadata []MetadataPair
+	Data     interface{}
+}
+
+func (request Request) MarshalJSON() ([]byte, error) {
+	var encFields []string
+
+	if request.Timeout.Seconds() != 0 {
+		encTimeout := fmt.Sprintf("\"timeoutSeconds\": %f", request.Timeout.Seconds())
+		encFields = append(encFields, encTimeout)
+	}
+
+	if len(request.Metadata) > 0 {
+		marshalMetadata, err := json.Marshal(request.Metadata)
+		if err != nil {
+			return nil, err
+		}
+		encMetadata := fmt.Sprintf("\"metadata\": %s", marshalMetadata)
+		encFields = append(encFields, encMetadata)
+	}
+
+	if request.Data != nil {
+		marshalData, err := marshalData(request.Data)
+		if err != nil {
+			return nil, err
+		}
+		encData := fmt.Sprintf("\"data\": %s", marshalData)
+		encFields = append(encFields, encData)
+	}
+
+	buffer := bytes.NewBufferString("{")
+	buffer.WriteString(strings.Join(encFields, ","))
+	buffer.WriteString("}")
+
+	return buffer.Bytes(), nil
+}
+
+func marshalData(data interface{}) ([]byte, error) {
+	var marshalData []byte
+	var err error
+
+	if protoMsg, ok := data.(proto.Message); ok {
+		marshalData, err = toJSON(protoMsg)
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		marshalData, err = json.Marshal(data)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return marshalData, nil
+}
+
+func toJSON(msg proto.Message) ([]byte, error) {
+	jsm := jsonpb.Marshaler{EmitDefaults: true, OrigName: true}
+	var b bytes.Buffer
+	if err := jsm.Marshal(&b, msg); err == nil {
+		return nil, err
+	}
+	return b.Bytes(), nil
+}
+
+func (request *Request) UnmarshalJSON(b []byte) (returnedErr error) {
+	var raw map[string]interface{}
```

Comment:
Instead of this very manual unmarshal, this could be improved in a couple of ways:

1. As mentioned above, if you change `Timeout time.Duration` to `TimeoutSeconds float32`, then you don't even need a custom unmarshal function at all. (I know I suggested `time.Duration`; I usually prefer type-safety. But I offer undoing that suggestion since I think the resulting code would be simpler).
2. OR, if you like type safety (who doesn't, right?), we could leave it `time.Duration`, but then use the technique described above for `Marshal`, where you unmarshal into a struct that has the exact right JSON format, and then convert from there. That way, you are only converting/adapting the field(s) that need it instead of having to write a map unmarshaller like this.